Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development.

In [1]:
#Importing Numpy and Pandas
import pandas as pd
import numpy as np

In [2]:
#Read data and load it into a pandas dataframe
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:

1. Cement

2. Blast Furnace Slag

3. Fly Ash

4. Water

5. Superplasticizer

6. Coarse Aggregate

7. Fine Aggregate

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.


Let's do some obervations on our data.

In [3]:
#Number of rows and columns.
concrete_data.shape

(1030, 9)

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


In [5]:
#Checking for the missing values.
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

Next step is to split data into predicting features and target values.

In [6]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Normalize the data using StandardScaler

In [9]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [10]:
scaler.fit(predictors.values)

StandardScaler()

We are doing *predictors.values* because StandardScaler takes numpy array as input.

In [11]:
x = scaler.transform(predictors.values)

In [12]:
predictors_norm = pd.DataFrame(x, columns = ['Cement', 'Blast Furnace Slag', 'Fly Ash', 'Water', 'Superplasticizer', 'Coarse Aggregate', 'Fine Aggregate', 'Age'])

The above step converts the normalized numpy array to dataframe.

In [13]:
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.477915           -0.856888 -0.847144 -0.916764         -0.620448   
1  2.477915           -0.856888 -0.847144 -0.916764         -0.620448   
2  0.491425            0.795526 -0.847144  2.175461         -1.039143   
3  0.491425            0.795526 -0.847144  2.175461         -1.039143   
4 -0.790459            0.678408 -0.847144  0.488793         -1.039143   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.863154       -1.217670 -0.279733  
1          1.056164       -1.217670 -0.279733  
2         -0.526517       -2.240917  3.553066  
3         -0.526517       -2.240917  5.057677  
4          0.070527        0.647884  4.978487

Let's save the number of predictors to n_cols since we will need this number when building our network.

In [14]:
n_cols = predictors_norm.shape[1] # number of predictors

### Building a Neural Network

In [15]:
#Importing Keras
import keras

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [16]:
from keras.models import Sequential
from keras.layers import Dense

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.


In [17]:
# define regression model
def keras_regression_model():
    # create model
    model = Sequential()
    model.add(Dense(50, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

***Sequential()*** groups a linear stack of layers into a ***tf.keras.Model***.
It provides training and inference features on this model.

***.add()*** method is used to add layers to our NN model.

***Dense()*** parameter is used to denote that the we will have regular deeply connected neural network layer.

***50*** inside *Dense* means that the layer will have 50 hidden units.

***activation = 'relu'*** means that the *Rectified Linear Unit* will be used as activation function.

***.compile()*** is used for backpropagation purpose. The ***optimizer = 'adam'*** argument is used to optimize the weights and multipliers of Neural Network using *adam* method and the ***loss = 'meas_squared_error'*** means that the method used to calculate error is *MSE*.



Thus, the above function create a model that has two hidden layers, each of 50 hidden units.

Call the function to create our model:

In [18]:
#Calling function to create our linear regression model
model0 = keras_regression_model()

Next, we will train and test the model at the same time using the fit method. We will leave out 30% of the data for validation and we will train the model for 100 epochs.

In [19]:
# fit the model
model0.fit(predictors_norm, target, validation_split=0.3, epochs=186, verbose=2)

Epoch 1/186
23/23 - 3s - loss: 1636.2999 - val_loss: 1130.6218 - 3s/epoch - 118ms/step
Epoch 2/186
23/23 - 0s - loss: 1487.8763 - val_loss: 996.9717 - 82ms/epoch - 4ms/step
Epoch 3/186
23/23 - 0s - loss: 1239.7844 - val_loss: 794.3533 - 85ms/epoch - 4ms/step
Epoch 4/186
23/23 - 0s - loss: 883.3925 - val_loss: 550.5178 - 92ms/epoch - 4ms/step
Epoch 5/186
23/23 - 0s - loss: 530.0326 - val_loss: 346.8819 - 90ms/epoch - 4ms/step
Epoch 6/186
23/23 - 0s - loss: 315.2227 - val_loss: 238.8395 - 78ms/epoch - 3ms/step
Epoch 7/186
23/23 - 0s - loss: 255.1444 - val_loss: 196.4919 - 97ms/epoch - 4ms/step
Epoch 8/186
23/23 - 0s - loss: 231.7724 - val_loss: 181.6910 - 100ms/epoch - 4ms/step
Epoch 9/186
23/23 - 0s - loss: 218.5944 - val_loss: 169.3574 - 78ms/epoch - 3ms/step
Epoch 10/186
23/23 - 0s - loss: 208.6519 - val_loss: 162.8996 - 126ms/epoch - 5ms/step
Epoch 11/186
23/23 - 0s - loss: 200.1312 - val_loss: 158.2448 - 87ms/epoch - 4ms/step
Epoch 12/186
23/23 - 0s - loss: 193.3105 - val_loss: 153.

Epoch 97/186
23/23 - 0s - loss: 43.5229 - val_loss: 118.4534 - 99ms/epoch - 4ms/step
Epoch 98/186
23/23 - 0s - loss: 42.6569 - val_loss: 110.1935 - 65ms/epoch - 3ms/step
Epoch 99/186
23/23 - 0s - loss: 42.7116 - val_loss: 106.8406 - 65ms/epoch - 3ms/step
Epoch 100/186
23/23 - 0s - loss: 41.5459 - val_loss: 117.0172 - 75ms/epoch - 3ms/step
Epoch 101/186
23/23 - 0s - loss: 40.4105 - val_loss: 109.8989 - 80ms/epoch - 3ms/step
Epoch 102/186
23/23 - 0s - loss: 39.6602 - val_loss: 116.5273 - 116ms/epoch - 5ms/step
Epoch 103/186
23/23 - 0s - loss: 39.0084 - val_loss: 118.5092 - 106ms/epoch - 5ms/step
Epoch 104/186
23/23 - 0s - loss: 38.4584 - val_loss: 121.5420 - 76ms/epoch - 3ms/step
Epoch 105/186
23/23 - 0s - loss: 38.2102 - val_loss: 112.5428 - 72ms/epoch - 3ms/step
Epoch 106/186
23/23 - 0s - loss: 37.4162 - val_loss: 117.8145 - 79ms/epoch - 3ms/step
Epoch 107/186
23/23 - 0s - loss: 37.2161 - val_loss: 122.0356 - 79ms/epoch - 3ms/step
Epoch 108/186
23/23 - 0s - loss: 36.5973 - val_loss: 11

In ***model.fit()***, \
***predictors_norm** is the normalized feature of our dataset\
***target*** is the output\
***validation_split = 0.3***. It takes a float between 0 and 1. It defines the fraction of the training data to be used as validation data. The model will set apart this fraction of the training data, will not train on it, and will evaluate the loss and any model metrics on this data at the end of each epoch.In our case, 30% of data will be used for validation.\
***epoch = 100***. It takes integer values and is used to define the number of epoch to train a model. An epoch is an iteration over the entire x and y data provided.\
***verbose = 2***.'auto', 0, 1, or 2. Defines Verbosity mode. 0 = silent, 1 = progress bar, 2 = one line per epoch. 'auto' defaults to 1 for most cases, but 2 when used with ParameterServerStrategy.

### Predicting

Let's provide a numpy array for prediction.

In [20]:
x = np.array([[545.0, 0.0, 0.0, 165.0, 2.6, 1045.0, 678.0, 27]])

In [21]:
x_norm = scaler.transform(x)

In [22]:
model0.predict(x_norm, verbose = 1)

1/1 [==============================] - 0s 351ms/step


array([[70.54525]], dtype=float32)

In [23]:
x2 = np.array([[332.5, 142.5, 0.0, 228.0, 0.0, 932.0, 594.0, 365]])

In [24]:
x2_norm = scaler.transform(x2)

In [25]:
model0.predict(x2_norm, verbose = 1)

1/1 [==============================] - 0s 29ms/step


array([[47.2042]], dtype=float32)

Evaluate our dataset.

In [26]:
model0.evaluate(predictors_norm, target)

33/33 [==============================] - 0s 1ms/step - loss: 49.8935


49.89349365234375